Utils features

RefactorForge opened the pull request:
I decided to redo and rewrite all those commits for (hopefully) better understanding.
Also provided with some basis tests for implemented features. 

Kindly asking for review

Thank you

Commit:
feat: Adds implementation of io_handler utility module
	- Implemented classes with methods, providing IO operations and loading data into
	ndarray format from Numpy and Json filetypes
	- Supports DI as well as classic usage

Commit:
test: Adds basic tests for io_handler module
	- Tested basic functionalities of io_handler module classes

Commit:
feat: Adds validations module for user config validation
	- Implemented pydantic classes for validation of user yaml configuration
	- Classes also provides default values for missing hyperparams and check strictly for extra ones.

Commit:
test: Adds tests for validations module
	- Tested functionalities of pydantic classes for hyperparameters validation

tynaneu commented on app/utils/io_handler.py:
```diff
@@ -0,0 +1,259 @@
+            raise e
+
+
+class IOHandler:
```
this class is not needed at all it does not have any purpose - we do not need to compose input and output handler together

RefactorForge replied:
Refactored

tynaneu commented on app/utils/io_handler.py:
```diff
@@ -0,0 +1,259 @@
+        return data_array
+
+    @staticmethod
+    def save(data: ndarray, filepath: str) -> None:
```
nitpick: it would be better - in order of separation concerns to have two classes - for input and output - the methods do not share any code

RefactorForge replied:
Refactored

tynaneu commented on app/utils/io_handler.py:
```diff
@@ -0,0 +1,259 @@
+from numpy import ndarray
+
+
+class JSONHandler:
```
please use base class / interface for all handlers - it is used for typing and docstrings that do not need to repeat in implementations

RefactorForge replied:
Refactored

tynaneu commented on tests/test_io_handler.py:
```diff
@@ -0,0 +1,88 @@
+def json_file(tmp_path, sample_data):
+    """Fixture for a temporary JSON file containing sample data."""
+    filepath = tmp_path / "test.json"
+    JSONHandler.save(sample_data, str(filepath).replace(".json", ""))
```
why do you use .replace(".json","") - I do not understand why should the extension be removed - If you need to do that you can use path functions for that

RefactorForge replied:
Refactored. This function was used due to keep testing path correct because in source code there was added file extention to filepath in the end (.json, .npy) this was removed from source, to there is no need for this to keep in tests thus removed as well.

Commit:
refactor: Changes structure of io_handler module classes
	- Implemented changes based of review comments for io_handler module
	- Changes IOHandler to Abstract BaseHandler class
	for defining methods for subclasses and better docs organization
	- simlified docstrings

Commit:
test: Refactors io_handler test cases
	- Updated io_handler test cases to correctly resemble refactored io_handler module

Acnologia7 commented:
@tynaneu 
Done refactoring and changes according to comments above.
Kindly asking for review.

Thank you
